### PR TITLE
Implement :visited support for lazy pseudo-elements

### PR DESCRIPTION
**Pull request by bzbarsky**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix https://bugzilla.mozilla.org/show_bug.cgi?id=1364242

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17570)

### CODE REVIEW

**Comment by bzbarsky** [comment]
@jryans care to review?

File: components/style/matching.rs [hunk 1]
```diff
@@ -288,17 +291,43 @@ trait PrivateMatchMethods: TElement {
         }
     }
 
+    /// Get the ComputedValues (if any) for our inheritance parent.
+    fn get_inherited_values_and_parent(&self) -> ParentElementAndValues<Self> {
```
Comment (jryans) [inline components/style/matching.rs]: Even though the type is `ComputedValues`, we tend to use the name `style` when talking about it in the code currently, so I'd say call this `get_inherited_style_and_parent` and the type `ParentElementAndStyle` with a `style` field for the CV type.

File: components/style/matching.rs [hunk 2]
```diff
@@ -288,17 +291,43 @@ trait PrivateMatchMethods: TElement {
...
+        }
+    }
+
     /// A common path for the cascade used by both primary elements and eager
     /// pseudo-elements after collecting the appropriate rules to use.
     ///
     /// `primary_style` is expected to be Some for eager pseudo-elements.
+    ///
+    /// `parent_info` is our style parent and its primary style, if
+    /// it's alreay been computed.
```
Comment (jryans) [inline components/style/matching.rs]: Nit: "already"

File: components/style/matching.rs [hunk 3]
```diff
@@ -391,11 +415,15 @@ trait PrivateMatchMethods: TElement {
     /// pseudo-elements.
     ///
     /// `primary_style` is expected to be Some for eager pseudo-elements.
+    ///
+    /// `parent_info` is our style parent and its primary style, if
+    /// it's alreay been computed.
```
Comment (jryans) [inline components/style/matching.rs]: Nit: "already"

File: components/style/matching.rs [hunk 4]
```diff
@@ -318,23 +347,18 @@ trait PrivateMatchMethods: TElement {
...
-                    // but not wanting to flush all of layout).
-                    debug_assert!(cfg!(feature = "gecko") ||
-                                  parent_el.unwrap().has_current_styles(d));
-                    d.styles.primary()
-                });
-                parent_style.map(|s| cascade_visited.values(s))
+                let info = match parent_info {
+                    Some(element_and_values) => element_and_values,
+                    None => {
+                        element_and_values = self.get_inherited_values_and_parent();
```
Comment (jryans) [inline components/style/matching.rs]: Do we actually end up going down this path? Or does every case where we need this data pass it down? (If it does, can we just `unwrap()` the option here?)

File: components/style/stylist.rs [hunk 5]
```diff
@@ -680,46 +681,89 @@ impl Stylist {
...
+            None
+        };
+
+        // We may not have non-visited rules, if we only had visited ones.  In
+        // that case we want to use the root rulenode for our non-visited rules.
+        let root;
+        let rules = if let Some(rules) = inputs.get_rules() {
+            rules
+        } else {
+            root = self.rule_tree.root();
```
Comment (jryans) [inline components/style/stylist.rs]: You could also make `rule_tree.root()` return a reference to avoid some of this.

**Review by jryans — CHANGES_REQUESTED** [review]
The logic generally seems sound, thanks for working on this! 😁 @emilio may also sprinkle in a few drive-by review comments.

File: components/style/stylist.rs (hunk 5 above)
Comment (bzbarsky) [inline components/style/stylist.rs]: Yeah, I wasn't sure whether it was purposefully returning a clone, so left it as-is.

File: components/style/matching.rs (hunk 1 above)
Comment (bzbarsky) [inline components/style/matching.rs]: Done.

File: components/style/matching.rs (hunk 2 above)
Comment (bzbarsky) [inline components/style/matching.rs]: Done.

File: components/style/matching.rs (hunk 3 above)
Comment (bzbarsky) [inline components/style/matching.rs]: Done.

File: components/style/matching.rs (hunk 4 above)
Comment (bzbarsky) [inline components/style/matching.rs]: We do end up going down this path, yes.  For example, get_after_change_style calls us with `CasacadeTarget::Normal` and a None for the `parent_info`.

**Pull request merged**